src/DomainHost.sol

```diff
@@ -0,0 +1,430 @@
+    function _undoDeposit(address originalSender, uint256 amount) internal {
+        require(dai.transferFrom(escrow, originalSender, amount), "DomainHost/transfer-failed");
+
+        emit UndoDeposit(originalSender, amount);
+    }
+
+    /// @notice Withdraw local DAI by burning remote canonical DAI
+    /// @param to The address to send the DAI to on the local domain
+    /// @param amount The amount of DAI to withdraw [WAD]
+    function withdraw(address to, uint256 amount) external guestOnly {
+        require(dai.transferFrom(escrow, to, amount), "DomainHost/transfer-failed");
+
+        emit Withdraw(to, amount);
+    }
+
+    // --- MCD Support ---
+
+    /// @notice Set the global debt ceiling for the remote domain
+    /// @dev Please note that pre-mint DAI cannot be removed from the remote domain
+    /// until the remote domain signals that it is safe to do so
+    /// @param wad The new debt ceiling [WAD]
+    function _lift(uint256 wad) internal auth returns (bytes memory payload) {
+        require(vat.live() == 1, "DomainHost/vat-not-live");
+
+        uint256 rad = wad * RAY;
+        uint256 dlineWad;
+        int256 dline = _int256(rad) - _int256(line);
+
+        if (dline > 0) {
+            // We are issuing new pre-minted DAI
+            dlineWad = uint256(dline) / RAY;                    // No precision loss as line is always a multiple of RAY
+            vat.slip(ilk, address(this), int256(dlineWad));     // No need for conversion check as amt is under a RAY of full size
+            vat.frob(ilk, address(this), address(this), address(this), int256(dlineWad), int256(dlineWad));
+            daiJoin.exit(escrow, dlineWad);
+
+            grain += dlineWad;
+        }
+
+        line = rad;
+
+        payload = abi.encodeWithSelector(DomainGuestLike.lift.selector, rid++, wad);
+
+        emit Lift(wad);
+    }
+
+    /// @notice Withdraw pre-mint DAI from the remote domain
+    /// @param _lid Local ordering id
+    /// @param wad The amount of DAI to release [WAD]
+    function release(uint256 _lid, uint256 wad) external guestOnly ordered(_lid) {
+        require(vat.live() == 1, "DomainHost/vat-not-live");
+
+        int256 amt = -_int256(wad);
+
+        // Fix any permissionless repays that may have occurred
+        (uint256 ink, uint256 art) = vat.urns(ilk, address(this));
+        if (art < ink) {
+            address _vow = vow;
+            uint256 diff = ink - art;
+            vat.suck(_vow, _vow, diff * RAY); // This needs to be done to make sure we can deduct sin[vow] and vice in the next call
+            vat.grab(ilk, address(this), address(this), _vow, 0, _int256(diff));
+        }
+
+        require(dai.transferFrom(escrow, address(this), wad), "DomainHost/transfer-failed");
+        daiJoin.join(address(this), wad);
+        vat.frob(ilk, address(this), address(this), address(this), amt, amt);
+        vat.slip(ilk, address(this), amt);
+
+        grain -= wad;
+
+        emit Release(wad);
+    }
```

Comment:
The definition of `amt` seems a bit confusing, probably more because is being done few lines above the actual usage. What about something like this:
```suggestion
    function release(uint256 _lid, uint256 wad) external guestOnly ordered(_lid) {
        require(vat.live() == 1, "DomainHost/vat-not-live");

        // Fix any permissionless repays that may have occurred
        (uint256 ink, uint256 art) = vat.urns(ilk, address(this));
        if (art < ink) {
            address _vow = vow;
            uint256 diff = ink - art;
            vat.suck(_vow, _vow, diff * RAY); // This needs to be done to make sure we can deduct sin[vow] and vice in the next call
            vat.grab(ilk, address(this), address(this), _vow, 0, _int256(diff));
        }

        require(dai.transferFrom(escrow, address(this), wad), "DomainHost/transfer-failed");
        daiJoin.join(address(this), wad);
        vat.frob(ilk, address(this), address(this), address(this), -_int256(wad), -int256(wad));
        vat.slip(ilk, address(this), -int256(wad));

        grain -= wad;

        emit Release(wad);
    }
```

Or maybe just defining the variable above the usage.

Reply:
anyway it is a nit, feel free to ignore

Reply:
Moved it and renamed to `repay`. How's that?

Reply:
Looks good!